usvm-ts/src/main/kotlin/org/usvm/machine/expr/TsLocalResolver.kt

```diff
@@ -0,0 +1,199 @@
+        // TODO: consider 'types.allocate'
+        for (captured in type.closures) {
+            val resolvedCaptured = resolveLocal(scope, captured) ?: return null
+            val lValue = mkFieldLValue(resolvedCaptured.sort, obj, captured.name)
+            scope.doWithState {
+                memory.write(lValue, resolvedCaptured.cast(), guard = trueExpr)
+            }
+        }
+        scope.doWithState {
+            setClosureObject(local.name, obj)
+        }
+        return obj
+    }
+
+    val lValue = resolveLocalToLValue(scope, local) ?: return null
+    return scope.calcOnState { memory.read(lValue) }
+}
+
+fun TsContext.resolveLocalToLValue(
+    scope: TsStepScope,
+    local: EtsValue,
+): ULValue<*, *>? {
+    val currentMethod = scope.calcOnState { lastEnteredMethod }
+
+    if (currentMethod.name == DEFAULT_ARK_METHOD_NAME) {
+        // TODO
+    }
+
+    // Get local index
+    val idx = getLocalIdx(local, currentMethod)
+
+    // If local is found in the current method:
+    if (idx != null) {
+        val sort = scope.calcOnState {
+            getOrPutSortForLocal(idx) {
+                val type = local.tryGetKnownType(currentMethod)
+                typeToSort(type).let {
+                    if (it is TsUnresolvedSort) {
+                        addressSort
+                    } else {
+                        it
+                    }
+                }
+            }
+        }
+        return mkRegisterStackLValue(sort, idx)
+    }
+
+    // Local not found, either global or imported
+    val file = currentMethod.enclosingClass!!.declaringFile!!
+    val globals = file.getGlobals()
+
+    require(local is EtsLocal) {
+        "Only locals are supported here, but got ${local::class.java}: $local"
+    }
+
+    // If local is a global variable:
+    if (globals.any { it.name == local.name }) {
+        val dfltObject = scope.calcOnState { getDfltObject(file) }
+
+        // Initialize globals in `file` if necessary
+        val isGlobalsInitialized = scope.calcOnState { isGlobalsInitialized(file) }
+        if (!isGlobalsInitialized) {
+            logger.info { "Globals are not initialized for file: $file" }
+            scope.doWithState {
+                initializeGlobals(file)
+            }
+            return null
+        } else {
+            // TODO: handle methodResult
+            scope.doWithState {
+                if (methodResult is TsMethodResult.Success) {
+                    methodResult = TsMethodResult.NoCall
+                }
+            }
+        }
+
+        // Try to get the saved sort for this dflt object field
+        val savedSort = scope.calcOnState {
+            getSortForDfltObjectField(file, local.name)
+        }
+
+        if (savedSort == null) {
+            // No saved sort means this field was never assigned to, which is an error
+            logger.error { "Trying to read unassigned global variable: '$local' in $file" }
+            scope.assert(falseExpr)
```

Comment:
If this is an error, you should throw an error. Not simply remove the state that will be an issue for someone later